matchms/similarity/spectrum_similarity_functions.py

```diff
@@ -5,69 +5,111 @@
-                       mz_power=0.0, intensity_power=1.0):
+def collect_peak_pairs(spec1: numpy.ndarray, spec2: numpy.ndarray,
+                       tolerance: float, shift: float = 0, mz_power: float = 0.0,
+                       intensity_power: float = 1.0):
     # pylint: disable=too-many-arguments
     """Find matching pairs between two spectra.
 
     Args
     ----
-    spec1: numpy array
+    spec1:
         Spectrum peaks and intensities as numpy array.
-    spec2: numpy array
+    spec2:
         Spectrum peaks and intensities as numpy array.
-    tolerance : float
+    tolerance
         Peaks will be considered a match when <= tolerance appart.
-    shift : float, optional
+    shift
         Shift spectra peaks by shift. The default is 0.
-    mz_power: float, optional
+    mz_power:
         The power to raise mz to in the cosine function. The default is 0, in which
         case the peak intensity products will not depend on the m/z ratios.
-    intensity_power: float, optional
+    intensity_power:
         The power to raise intensity to in the cosine function. The default is 1.
 
     Returns
     -------
-    matching_pairs : list
-        List of found matching peaks.
+    matching_pairs : numpy array
+        Array of found matching peaks.
     """
+    matches = find_matches(spec1, spec2, tolerance, shift)
+    idx1 = [x[0] for x in matches]
+    idx2 = [x[1] for x in matches]
+    if len(idx1) == 0:
+        return None
     matching_pairs = []
+    for i, idx in enumerate(idx1):
+        power_prod_spec1 = (spec1[idx, 0] ** mz_power) * (spec1[idx, 1] ** intensity_power)
+        power_prod_spec2 = (spec2[idx2[i], 0] ** mz_power) * (spec2[idx2[i], 1] ** intensity_power)
+        matching_pairs.append([idx1[i], idx2[i], power_prod_spec1 * power_prod_spec2])
+    return matching_pairs
 
-    for idx in range(len(spec1)):
-        intensity = spec1[idx, 1]
-        mz = spec1[idx, 0]
-        matches = numpy.where((numpy.abs(spec2[:, 0] - spec1[idx, 0] + shift) <= tolerance))[0]
-        for match in matches:
-            power_prod_spec1 = ((mz ** mz_power) * (intensity ** intensity_power))
-            power_prod_spec2 = ((spec2[match][0] ** mz_power) * (spec2[match][1] ** intensity_power))
-            matching_pairs.append((idx, match, power_prod_spec1 * power_prod_spec2))
 
-    return matching_pairs
+@numba.njit
+def find_matches(spec1, spec2, tolerance, shift):
+    """Faster search for matching peaks.
+    Makes use of the fact that spec1 and spec2 contain ordered peak m/z (from
+    low to high m/z).
+
+    Parameters
+    ----------
+    spec1:
+        Spectrum peaks and intensities as numpy array. Peak mz values must be ordered.
+    spec2:
+        Spectrum peaks and intensities as numpy array. Peak mz values must be ordered.
+    tolerance
+        Peaks will be considered a match when <= tolerance appart.
+    shift
+        Shift spectra peaks by shift. The default is 0.
```

Comment:
On line 78 you are shifting only the peaks in the second spectra. Can the docs reflect this?
```suggestion
        Shift peaks of second spectra by shift. The default is 0.
```

Reply:
Good point, was added.